odl/space/fspace.py

```diff
@@ -15,581 +15,233 @@
+    is_real_dtype, is_complex_floating_dtype, dtype_repr, dtype_str,
+    complex_dtype, real_dtype, signature_string,
     is_valid_input_array, is_valid_input_meshgrid,
-    out_shape_from_array, out_shape_from_meshgrid, vectorize)
+    out_shape_from_array, out_shape_from_meshgrid, vectorize, broadcast_to,
+    writable_array)
 from odl.util.utility import preload_first_arg
 
 
-__all__ = ('FunctionSet', 'FunctionSetElement',
-           'FunctionSpace', 'FunctionSpaceElement')
+__all__ = ('FunctionSpace',)
+
+
+def _check_out_arg(func):
+    """Check if of ``func`` has an (optional) ``out`` argument.
+
+    Also verify that the signature of ``func`` has no ``*args`` since
+    they make argument propagation a hassle.
+
+    Parameters
+    ----------
+    func : callable
+        Object that should be inspected.
+
+    Returns
+    -------
+    has_out : bool
+        ``True`` if the signature has an ``out`` argument, ``False``
+        otherwise.
+    out_is_optional : bool
+        ``True`` if ``out`` is present and optional in the signature,
+        ``False`` otherwise.
+
+    Raises
+    ------
+    TypeError
+        If ``func``'s signature has ``*args``.
+    """
+    if sys.version_info.major > 2:
+        spec = inspect.getfullargspec(func)
+        kw_only = spec.kwonlyargs
+    else:
+        spec = inspect.getargspec(func)
+        kw_only = ()
+
+    if spec.varargs is not None:
+        raise TypeError('*args not allowed in function signature')
+
+    pos_args = spec.args
+    pos_defaults = () if spec.defaults is None else spec.defaults
+
+    has_out = 'out' in pos_args or 'out' in kw_only
+    if 'out' in pos_args:
+        has_out = True
+        out_is_optional = (
+            pos_args.index('out') >= len(pos_args) - len(pos_defaults))
+    elif 'out' in kw_only:
+        has_out = out_is_optional = True
+    else:
+        has_out = out_is_optional = False
+
+    return has_out, out_is_optional
 
 
 def _default_in_place(func, x, out, **kwargs):
     """Default in-place evaluation method."""
-    out[:] = func(x, **kwargs)
+    result = np.array(func._call_out_of_place(x, **kwargs), copy=False)
+    if result.dtype == object:
+        # Different shapes encountered, need to broadcast
+        flat_results = result.ravel()
+        if is_valid_input_array(x, func.domain.ndim):
+            scalar_out_shape = out_shape_from_array(x)
+        elif is_valid_input_meshgrid(x, func.domain.ndim):
+            scalar_out_shape = out_shape_from_meshgrid(x)
+        else:
+            raise RuntimeError('bad input')
+
+        bcast_results = [broadcast_to(res, scalar_out_shape)
+                         for res in flat_results]
+        result = np.array(bcast_results, dtype=func.scalar_out_dtype)
+        result = result.reshape(func.out_shape + scalar_out_shape)
+
+    try:
+        reshaped = result.reshape(out.shape)
+    except ValueError:
+        out[:] = result
+    else:
+        out[:] = reshaped
```

Comment:
This was the simplest way to handle all annoying cases with extra dimensions. Not sure if this should be handled more explicitly.